feat(cli,sdk): qwen serve daemon (Stage 1) (#3889)
QwenLM/qwen-code: docs/developers/qwen-serve-protocol.md

When the daemon was started with `--token` or `QWEN_SERVER_TOKEN`, every request must carry:

```
Authorization: Bearer <token>
```

Without a configured token (loopback dev default) the header is optional. Token comparison is constant-time. 401 responses are uniform across `missing header` / `wrong scheme` / `wrong token`.

## Common error shape

5xx responses carry the original error's `code` and `data` when present (JSON-RPC style — the ACP SDK forwards `{code, message, data}` from the agent):

```json
{
  "error": "Internal error",
  "code": -32000,
  "data": { "reason": "model quota exceeded" }
}
```

Malformed JSON in a request body returns:

```json
{ "error": "Invalid JSON in request body" }
```

with status `400`.

`SessionNotFoundError` for an unknown session id returns:

```json
{ "error": "No session with id \"<sid>\"", "sessionId": "<sid>" }
```

with status `404`.

## Capabilities

Every Stage 1 daemon advertises 9 feature tags. Clients **must** gate UI off `features`, not off `mode` (per design §10).

```
['health', 'capabilities', 'session_create', 'session_list',
 'session_prompt', 'session_cancel', 'session_events',
 'session_set_model', 'permission_vote']
```

> ## Routes

Author: **[Critical]** No `DELETE /session/:id` route exists. The multi-prompt contract mentions "kill the session via channel exit" but HTTP clients have no session-close endpoint. Sessions live until daemon exit, with no documented lifecycle management.

```suggestion
> **Stage 1 limitation:** Sessions live until daemon exit; there is no HTTP session-close route. `DELETE /session/:id` is planned for Stage 2.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

### `GET /health`

Liveness probe. Returns `200 {"status":"ok"}` if the listener is up. No auth required even when a token is configured (heartbeat-friendly).

> ### `GET /capabilities`

Author: **[Suggestion]** `GET /capabilities` always shows `modelServices: []` but `POST /session` accepts `modelServiceId`. If it's always empty in Stage 1, document that explicitly so client developers know to omit `modelServiceId`.

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

```json
{
  "v": 1,
  "mode": "http-bridge",
  "features": ["health", "capabilities", "..."],
  "modelServices": []
}
```

Stable contract: when `v` increments the frame layout has changed in a backwards-incompatible way.

> ### `POST /session`

Author: **[Suggestion]** Concurrent `POST /session` coalescing behavior doesn't document the failure path — if the single spawn fails, all coalesced callers receive the same error. This should be documented as part of the protocol contract.

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

Request:

```json
{
  "cwd": "/absolute/path/to/workspace",
  "modelServiceId": "qwen-prod"
}
```

| Field            | Required | Notes                                                                                                                                                                                                                           |
| ---------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `cwd`            | yes      | Absolute path. Relative paths return `400`. Workspace paths are canonicalized via `realpathSync.native` (with a resolve-only fallback for non-existent paths) so case-insensitive filesystems don't fork sessions per spelling. |
| `modelServiceId` | no       | If omitted the agent uses its default model. If the workspace already has a session, this calls `setSessionModel` on the existing one and broadcasts `model_switched`.                                                          |

Response:

> ```json

Author: **[Suggestion]** `POST /session` uses `modelServiceId` but `POST /session/:id/model` uses `modelId`. These appear to be the same identifier — consider unifying the field name across both endpoints for consistency.

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

: heartbeat              ← every 15s, no payload

event: client_evicted    ← terminal frame, no id (synthetic)
data: {"reason":"queue_overflow","droppedAfter":42}
```

| Event type            | Trigger                                                              |
| --------------------- | -------------------------------------------------------------------- |
| `session_update`      | Any ACP `sessionUpdate` notification (LLM chunks, tool calls, usage) |
| `permission_request`  | Agent asked for tool approval                                        |
| `permission_resolved` | Some client voted on a permission via `POST /permission/:requestId`  |
| `model_switched`      | `POST /session/:id/model` succeeded                                  |
| `model_switch_failed` | `POST /session/:id/model` rejected                                   |
| `session_died`        | Agent child crashed unexpectedly                                     |
| `client_evicted`      | Subscriber-local: queue overflow (terminal, no id)                   |
| `stream_error`        | Daemon-side error during fan-out (terminal, no id)                   |

> Reconnect semantics:

Author: **[Critical]** Ring buffer reconnect can silently drop events. When `Last-Event-ID` predates the oldest event in the 1000-depth ring, the daemon replays from the oldest available event without indicating the gap. Clients may operate on incomplete state without knowing they missed events.

```suggestion
If the requested `Last-Event-ID` falls outside the ring buffer window, the daemon replays from the oldest available event without indicating loss. Clients should compare the first replayed event's `id` against their expected cursor to detect gaps.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

- Send `Last-Event-ID: <n>` to replay events with `id > n` from the per-session ring (default depth 1000)
- IDs are monotonic per session, starting at 1
- Synthetic terminal frames (`client_evicted`, `stream_error`) intentionally omit `id` so they don't burn a sequence slot for other subscribers

Backpressure:

- Per-subscriber queue defaults to `maxQueued: 256` live items (replay frames during reconnect bypass the cap)
- On overflow the bus emits the `client_evicted` terminal frame and closes the subscription

> ### `POST /permission/:requestId`

Author: **[Critical]** Permission requests have no timeout — the agent hangs indefinitely if no client votes. The `requestPermission` promise only resolves on vote, session cancel, or daemon shutdown. In headless/CI mode with disconnected clients, this causes silent hangs.

```suggestion
**Note:** If no client responds, the permission request remains pending indefinitely (until session cancel or daemon shutdown). Stage 2 will add a configurable timeout.
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

After a successful vote, every connected client sees `permission_resolved` with the same `requestId` and the chosen `outcome`.

## Streaming wire format

Events are emitted as standard EventSource frames. The daemon writes one `data:` line per frame (the JSON has no embedded newlines after `JSON.stringify`); the SDK parser at `packages/sdk-typescript/src/daemon/sse.ts` handles both that and the spec-allowed multi-`data:` form on the receive side.

## Error frames during streaming

If the bridge iterator throws while serving an SSE subscriber, the daemon emits a terminal `stream_error` frame (no `id`):

```
event: stream_error
data: {"error":"<message>"}
```

The connection then closes.

## Environment variables

| Var                 | Purpose                                                                                                      |
| ------------------- | ------------------------------------------------------------------------------------------------------------ |
| `QWEN_SERVER_TOKEN` | Bearer token. Stripped of leading/trailing whitespace at boot.                                               |

> | `QWEN_E2E_LLM`      | Set to `1` to enable LLM-required integration tests in `integration-tests/cli/qwen-serve-streaming.test.ts`. |

Author: **[Suggestion]** Environment variable `QWEN_E2E_LLM` doesn't exist in the codebase. The actual variable is `SKIP_LLM_TESTS` (inverted logic). Setting `QWEN_E2E_LLM=1` does nothing.

```suggestion
| `SKIP_LLM_TESTS`    | Set to `1` to skip LLM-required integration tests in `integration-tests/cli/qwen-serve-streaming.test.ts`. |
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_

| ---------------------------------------------------- | -------------------------------------------------------- |
| `packages/cli/src/commands/serve.ts`                 | yargs command + flag schema                              |
| `packages/cli/src/serve/runQwenServe.ts`             | listener lifecycle + signal handling                     |
| `packages/cli/src/serve/server.ts`                   | Express routes + middleware                              |
| `packages/cli/src/serve/auth.ts`                     | bearer + Host allowlist + CORS deny                      |
| `packages/cli/src/serve/httpAcpBridge.ts`            | spawn-or-attach + per-session FIFO + permission registry |
| `packages/cli/src/serve/eventBus.ts`                 | bounded async queue + replay ring                        |
| `packages/sdk-typescript/src/daemon/DaemonClient.ts` | TS client                                                |
| `packages/sdk-typescript/src/daemon/sse.ts`          | EventSource frame parser                                 |
| `integration-tests/cli/qwen-serve-routes.test.ts`    | 18 cases, no LLM                                         |

> | `integration-tests/cli/qwen-serve-streaming.test.ts` | 4 cases, real `qwen --acp` child (`QWEN_E2E_LLM=1`)      |

Author: **[Suggestion]** Wrong test count: `qwen-serve-streaming.test.ts` has 3 `it()` blocks, not 4.

```suggestion
| `integration-tests/cli/qwen-serve-streaming.test.ts` | 3 cases, real `qwen --acp` child (`SKIP_LLM_TESTS=1` skips) |
```

_— DeepSeek/deepseek-v4-pro via Qwen Code /review_
